test/sql/join/asof/test_asof_join.test

```diff
@@ -0,0 +1,352 @@
+# name: test/sql/join/asof/test_asof_join.test
+# description: Test As-Of joins
+# group: [asof]
+
+statement ok
+PRAGMA enable_verification
+
+#
+# Inequality only
+#
+
+# Use doubles for readable infinities
+statement ok
+CREATE TABLE events0 (begin DOUBLE, value INTEGER);
+
+statement ok
+INSERT INTO events0 VALUES
+	(1, 0),
+	(3, 1),
+	(6, 2),
+	(8, 3)
+;
+
+# INNER Window version
+query II nosort innert_inequality
+SELECT p.ts, e.value
+FROM 
+	range(0,10) p(ts)
+JOIN (
+	SELECT value, begin, 
+		LEAD(begin, 1, 'infinity'::DOUBLE) OVER (ORDER BY begin ASC) AS end
+	FROM events0
+) e
+ON p.ts >= e.begin AND p.ts < e.end
```

Comment:
Could we add some extra tests:

* Join on a timestamp range
* Join on a string range (it is fine if that throws a non-internal exception)
* Join on integer ranges (it is fine if that throws a non-internal exception)
* Multiple ranges in one join condition
* What if we invert the comparisons (e.g. `e.begin <= p.ts AND e.end > p.ts`)
* What if we have a different closed range (e.g. `p.ts > e.begin AND p.ts <= e.end`)
* What if we have a range that is not actually a range (e.g. `p.ts >= e.begin and p.other_col < e.end`)
* ASOF join on a constant (e.g. `ON 1=1`)
* Could we add some tests where we have NULL values in the tables?
* Could we add some tests where we have infinity values (e.g. `NaN`, or `infinity::timestamp`) in the tables?
* What happens if we use an ASOF join inside of a correlated subquery? (it is fine if that throws an exception)

Reply:
Multiple ranges was covered by the last error test, and you can't have invalid ranges because the bounds are implied (i.e., `end` is synthesised). But I'll have a go at the rest.

Reply:
Got all of these except the last one. How would you write an ASOF join inside a correlated subquery? Aren't the joins in correlated subqueries generated by the planner?

Reply:
Ah, I meant an ASOF join inside a subquery that has correlated columns in the left and/or right side, see e.g. [here](https://github.com/duckdb/duckdb/blob/master/test/sql/subquery/scalar/test_subquery_any_join.test#L4)